File: Min Stack.py
# https://leetcode.com/problems/min-stack/

import sys
class MinStack:

    def __init__(self):
        
        # Create a list
        self.stack = list()
        
        # Hold minimum value
        self.minElem = [sys.maxsize, -1]

    def push(self, val: int) -> None:
        
        # OBJECTIVE: Add new number to end of list and update minVal if possible
            
        # Add new value
        self.stack.append(val)
        
        # Update minVal
        if val <= self.minElem[0]:
            self.minElem = [val, len(self.stack) - 1]

    def pop(self) -> None:
        
        # OBJECTIVE: Remove last element from stack
        
        # Remove last value
        self.stack.pop(-1)
        
        # If stack is empty, reset minElem 
        if len(self.stack) == 0:
            self.minElem = [sys.maxsize, -1]
            
        # If stack is non-empty, check if minElem needs to be updated
        # E.g., The last element could have been the smallest one, so minElem needs to be updated
        elif self.minElem[1] >= len(self.stack) - 1:
            minVal = min(self.stack)
            self.minElem = [minVal, self.stack.index(minVal)]
        
    def top(self) -> int:
        
        # OBJECTIVE: Return last element from stack
        return self.stack[-1]

    def getMin(self) -> int:
        
        # OBJECTIVE: Return minimum element in stack
        return self.minElem[0]

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: Min Stack.java
// https://leetcode.com/problems/min-stack/
// https://youtu.be/qkLl7nAwDPo?si=_Z5hYCZI4qnlr7dI

/*
Time Complexity: O(1) for all operations because all stack operations guarantees O(1) time
Space Complexity: O(n) where n = size of mainStack and minStack (whichever is the largest).
*/

class MinStack {

    Stack<Integer> mainStack;
    Stack<Integer> minTrackerStack;

    public MinStack() {
        this.mainStack = new Stack<>();
        this.minTrackerStack = new Stack<>();
    }
    
    public void push(int val) {
        Integer valInteger = Integer.valueOf(val);

        if (mainStack.empty()) {
            mainStack.push(valInteger);
            minTrackerStack.push(valInteger);
        } else {
            mainStack.push(valInteger);

            Integer trackerPeekElem = minTrackerStack.peek();
            Integer smallestValue = valInteger > trackerPeekElem ? trackerPeekElem : valInteger;
            minTrackerStack.push(smallestValue);
        }
    }
    
    public void pop() {
        mainStack.pop();
        minTrackerStack.pop();
    }
    
    public int top() {
        return mainStack.peek().intValue();
    }
    
    public int getMin() {
        return minTrackerStack.peek().intValue();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */
